File: ios/PS Vault/PS Vault/Services/APIService.swift
import Foundation

enum APIError: Error, LocalizedError {
    case noServerURL
    case httpError(Int, String)
    case decodingError(Error)
    case networkError(Error)
    case unauthorized

    var errorDescription: String? {
        switch self {
        case .noServerURL: return "No server URL configured."
        case .httpError(_, let msg): return msg
        case .decodingError(let e): return "Decoding error: \(e)"
        case .networkError(let e): return e.localizedDescription
        case .unauthorized: return "Session expired. Please log in again."
        }
    }
}

@MainActor
final class APIService {
    static let shared = APIService()
    private init() {}

    var baseURL: String = ""
    var accessToken: String? = nil

    private let session = URLSession.shared
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    // MARK: - Core request

    private func buildRequest(_ method: String, path: String, body: (any Encodable)? = nil) throws -> URLRequest {
        guard !baseURL.isEmpty else { throw APIError.noServerURL }
        guard let url = URL(string: "\(baseURL)/api/v1\(path)") else {
            throw APIError.networkError(URLError(.badURL))
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("mobile", forHTTPHeaderField: "X-Client")
        if let token = accessToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder().encode(body)
        }
        return req
    }

    private func parseEnvelopeError(data: Data, statusCode: Int) -> APIError {
        if let envelope = try? decoder.decode(APIEnvelope<EmptyData>.self, from: data),
           let errorBody = envelope.error {
            if statusCode == 401 && errorBody.code == "unauthorized" && accessToken != nil {
                return .unauthorized
            }
            return .httpError(statusCode, errorBody.code)
        }
        return .httpError(statusCode, HTTPURLResponse.localizedString(forStatusCode: statusCode))
    }

    private func request<T: Decodable>(
        _ method: String,
        path: String,
        body: (any Encodable)? = nil
    ) async throws -> T {
        do {
            let req = try buildRequest(method, path: path, body: body)
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.networkError(URLError(.badServerResponse))
            }
            if !(200...299).contains(http.statusCode) {
                throw parseEnvelopeError(data: data, statusCode: http.statusCode)
            }
            let envelope = try decoder.decode(APIEnvelope<T>.self, from: data)
            if let result = envelope.data { return result }
            if let errorBody = envelope.error { throw APIError.httpError(http.statusCode, errorBody.code) }
            throw APIError.decodingError(DecodingError.dataCorrupted(.init(codingPath: [], debugDescription: "No data in envelope")))
        } catch let e as APIError {
            throw e
        } catch let e as DecodingError {
            throw APIError.decodingError(e)
        } catch {
            throw APIError.networkError(error)
        }
    }

    private func requestVoid(
        _ method: String,
        path: String,
        body: (any Encodable)? = nil
    ) async throws {
        do {
            let req = try buildRequest(method, path: path, body: body)
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.networkError(URLError(.badServerResponse))
            }
            if http.statusCode == 204 { return }
            if !(200...299).contains(http.statusCode) {
                throw parseEnvelopeError(data: data, statusCode: http.statusCode)
            }
        } catch let e as APIError {
            throw e
        } catch {
            throw APIError.networkError(error)
        }
    }

    // MARK: - Auth

    func login(email: String, password: String, mfaCode: String? = nil) async throws -> AuthResponse {
        struct Body: Encodable { let email, password: String; let mfa_code: String? }
        return try await request("POST", path: "/auth/login", body: Body(email: email, password: password, mfa_code: mfaCode))
    }

    func register(email: String, password: String, displayName: String, mekSalt: String, mekEnvelope: String) async throws -> AuthResponse {
        struct Body: Encodable { let email, password, display_name, mek_salt, mek_envelope: String }
        return try await request("POST", path: "/auth/register", body: Body(email: email, password: password, display_name: displayName, mek_salt: mekSalt, mek_envelope: mekEnvelope))
    }

    func refreshToken(_ token: String) async throws -> AuthResponse {
        struct Body: Encodable { let refresh_token: String }
        return try await request("POST", path: "/auth/refresh", body: Body(refresh_token: token))
    }

    func logout(refreshToken: String) async throws {
        struct Body: Encodable { let refresh_token: String }
        try await requestVoid("POST", path: "/auth/logout", body: Body(refresh_token: refreshToken))
    }

    func forgotPassword(email: String) async throws {
        struct Body: Encodable { let email: String }
        try await requestVoid("POST", path: "/auth/forgot-password", body: Body(email: email))
    }

    func resetPassword(token: String, password: String) async throws {
        struct Body: Encodable { let token, password: String }
        try await requestVoid("POST", path: "/auth/reset-password", body: Body(token: token, password: password))
    }

    // MARK: - User

    func getMe() async throws -> User {
        return try await request("GET", path: "/users/me")
    }

    func updateMe(displayName: String) async throws -> User {
        struct Body: Encodable { let display_name: String }
        return try await request("PATCH", path: "/users/me", body: Body(display_name: displayName))
    }

    func changeEmail(newEmail: String, currentPassword: String) async throws {
        struct Body: Encodable { let new_email, current_password: String }
        try await requestVoid("POST", path: "/users/me/change-email", body: Body(new_email: newEmail, current_password: currentPassword))
    }

    func changePassword(currentPassword: String, newPassword: String, newMEKEnvelope: String) async throws {
        struct Body: Encodable { let current_password, new_password, new_mek_envelope: String }
        try await requestVoid("POST", path: "/users/me/change-password", body: Body(current_password: currentPassword, new_password: newPassword, new_mek_envelope: newMEKEnvelope))
    }

    func registerPushToken(_ token: String, platform: String) async throws {
        struct Body: Encodable { let token, platform: String }
        try await requestVoid("POST", path: "/users/me/push-token", body: Body(token: token, platform: platform))
    }

    func deletePushToken() async throws {
        try await requestVoid("DELETE", path: "/users/me/push-token")
    }

    // MARK: - Sessions

    func listSessions() async throws -> [Session] {
        return try await request("GET", path: "/users/me/sessions")
    }

    func revokeSession(_ id: String) async throws {
        try await requestVoid("DELETE", path: "/users/me/sessions/\(id)")
    }

    func revokeAllSessions() async throws {
        try await requestVoid("DELETE", path: "/users/me/sessions")
    }

    // MARK: - Vaults

    func listVaults() async throws -> [Vault] {
        return try await request("GET", path: "/vaults")
    }

    func createVault(name: String, icon: String, cekEnvelope: String) async throws -> Vault {
        struct Body: Encodable { let name, icon, cek_envelope: String }
        return try await request("POST", path: "/vaults", body: Body(name: name, icon: icon, cek_envelope: cekEnvelope))
    }

    func updateVault(_ id: String, name: String? = nil, icon: String? = nil) async throws -> Vault {
        struct Body: Encodable { let name: String?; let icon: String? }
        return try await request("PATCH", path: "/vaults/\(id)", body: Body(name: name, icon: icon))
    }

    func deleteVault(_ id: String) async throws {
        try await requestVoid("DELETE", path: "/vaults/\(id)")
    }

    // MARK: - Entries

    func listEntries(vaultId: String) async throws -> [VaultEntry] {
        return try await request("GET", path: "/vaults/\(vaultId)/entries")
    }

    func createEntry(vaultId: String, entryType: String, title: String, encryptedData: String, sortOrder: Int? = nil) async throws -> VaultEntry {
        struct Body: Encodable { let entry_type, title, encrypted_data: String; let sort_order: Int? }
        return try await request("POST", path: "/vaults/\(vaultId)/entries", body: Body(entry_type: entryType, title: title, encrypted_data: encryptedData, sort_order: sortOrder))
    }

    func updateEntry(_ vaultId: String, entryId: String, title: String? = nil, encryptedData: String? = nil, sortOrder: Int? = nil, isFavorite: Bool? = nil) async throws -> VaultEntry {
        struct Body: Encodable { let title: String?; let encrypted_data: String?; let sort_order: Int?; let is_favorite: Bool? }
        return try await request("PATCH", path: "/vaults/\(vaultId)/entries/\(entryId)", body: Body(title: title, encrypted_data: encryptedData, sort_order: sortOrder, is_favorite: isFavorite))
    }

    func deleteEntry(_ vaultId: String, entryId: String) async throws {
        try await requestVoid("DELETE", path: "/vaults/\(vaultId)/entries/\(entryId)")
    }

    // MARK: - Vault Beneficiaries

    func getVaultBeneficiaries(vaultId: String) async throws -> [VaultBeneficiary] {
        return try await request("GET", path: "/vaults/\(vaultId)/beneficiaries")
    }

    func assignBeneficiary(vaultId: String, beneficiaryId: String, cekEnvelope: String) async throws {
        struct Body: Encodable { let beneficiary_id, beneficiary_cek_envelope: String }
        try await requestVoid("POST", path: "/vaults/\(vaultId)/beneficiaries", body: Body(beneficiary_id: beneficiaryId, beneficiary_cek_envelope: cekEnvelope))
    }

    func removeVaultBeneficiary(vaultId: String, beneficiaryId: String) async throws {
        try await requestVoid("DELETE", path: "/vaults/\(vaultId)/beneficiaries/\(beneficiaryId)")
    }

    // MARK: - Beneficiaries

    func listBeneficiaries() async throws -> [Beneficiary] {
        return try await request("GET", path: "/beneficiaries")
    }

    func createBeneficiary(name: String, email: String, relationship: String? = nil, secretQuestion: String? = nil, photoData: String? = nil) async throws -> Beneficiary {
        struct Body: Encodable { let name, email: String; let relationship: String?; let secret_question: String?; let photo_data: String? }
        return try await request("POST", path: "/beneficiaries", body: Body(name: name, email: email, relationship: relationship, secret_question: secretQuestion, photo_data: photoData))
    }

    func updateBeneficiary(_ id: String, name: String? = nil, relationship: String? = nil, secretQuestion: String? = nil, photoData: String? = nil) async throws -> Beneficiary {
        struct Body: Encodable { let name: String?; let relationship: String?; let secret_question: String?; let photo_data: String? }
        return try await request("PATCH", path: "/beneficiaries/\(id)", body: Body(name: name, relationship: relationship, secret_question: secretQuestion, photo_data: photoData))
    }

    func deleteBeneficiary(_ id: String) async throws {
        try await requestVoid("DELETE", path: "/beneficiaries/\(id)")
    }

    func resendBeneficiaryConfirmation(_ id: String) async throws {
        try await requestVoid("POST", path: "/beneficiaries/\(id)/resend")
    }

    // MARK: - Switch

    func getSwitchSettings() async throws -> SwitchSettings {
        return try await request("GET", path: "/switch")
    }

    func updateSwitchSettings(
        isActive: Bool? = nil,
        checkInIntervalDays: Int? = nil,
        abortWindowHours: Int? = nil,
        reminder1DaysBefore: Int? = nil,
        reminder2HoursBefore: Int? = nil,
        finalWarningHoursBefore: Int? = nil,
        preferredCheckinHour: Int? = nil,
        clearPreferredHour: Bool? = nil
    ) async throws -> SwitchSettings {
        struct Body: Encodable {
            let is_active: Bool?
            let check_in_interval_days: Int?
            let abort_window_hours: Int?
            let reminder1_days_before: Int?
            let reminder2_hours_before: Int?
            let final_warning_hours_before: Int?
            let preferred_checkin_hour: Int?
            let clear_preferred_hour: Bool?
        }
        return try await request("PATCH", path: "/switch", body: Body(
            is_active: isActive,
            check_in_interval_days: checkInIntervalDays,
            abort_window_hours: abortWindowHours,
            reminder1_days_before: reminder1DaysBefore,
            reminder2_hours_before: reminder2HoursBefore,
            final_warning_hours_before: finalWarningHoursBefore,
            preferred_checkin_hour: preferredCheckinHour,
            clear_preferred_hour: clearPreferredHour
        ))
    }

    func checkin() async throws -> SwitchSettings {
        return try await request("POST", path: "/switch/checkin")
    }

    func pauseSwitch(resumeAt: Date? = nil) async throws -> SwitchSettings {
        struct Body: Encodable { let resume_at: String? }
        let iso = resumeAt.map { ISO8601DateFormatter().string(from: $0) }
        return try await request("POST", path: "/switch/pause", body: Body(resume_at: iso))
    }

    func resumeSwitch() async throws -> SwitchSettings {
        return try await request("POST", path: "/switch/resume")
    }

    func abortTrigger() async throws -> SwitchSettings {
        return try await request("POST", path: "/switch/abort")
    }

    func revokeDeliveries() async throws {
        try await requestVoid("POST", path: "/switch/revoke-deliveries")
    }

    // MARK: - MFA

    func setupTOTP() async throws -> TOTPSetupResponse {
        return try await request("POST", path: "/auth/mfa/setup")
    }

    func confirmTOTP(secret: String, code: String, backupCodes: [String]) async throws {
        struct Body: Encodable { let secret, code: String; let backup_codes: [String] }
        try await requestVoid("POST", path: "/auth/mfa/verify", body: Body(secret: secret, code: code, backup_codes: backupCodes))
    }

    func disableMFA(code: String) async throws {
        struct Body: Encodable { let code: String }
        try await requestVoid("POST", path: "/auth/mfa/disable", body: Body(code: code))
    }

    // MARK: - Branding

    struct BrandingResponse: Decodable {
        let accentColor: String
        enum CodingKeys: String, CodingKey { case accentColor = "accent_color" }
    }

    func getBranding() async throws -> BrandingResponse {
        return try await request("GET", path: "/branding")
    }

    func updateAccentColor(_ hex: String) async throws {
        try await requestVoid("PATCH", path: "/admin/config", body: ["app_accent_color": hex])
    }
}

File: android/app/src/main/kotlin/dev/psvault/app/api/ApiService.kt
package dev.psvault.app.api

import android.os.Handler
import android.os.Looper
import dev.psvault.app.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

sealed class ApiException(message: String) : Exception(message) {
    class NoServerUrl : ApiException("No server URL configured.")
    class HttpError(val statusCode: Int, val errorCode: String) : ApiException(errorCode)
    class NetworkError(cause: Throwable) : ApiException(cause.message ?: "Network error")
    class DecodingError(cause: Throwable) : ApiException("Decoding error: ${cause.message}")
    object Unauthorized : ApiException("Session expired. Please log in again.")
}

object ApiService {

    var baseUrl: String = ""
    var accessToken: String? = null
    var onUnauthorized: (() -> Unit)? = null

    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()

    // MARK: - Core request

    private fun buildRequest(method: String, path: String, bodyJson: String? = null): Request {
        if (baseUrl.isEmpty()) throw ApiException.NoServerUrl()
        val url = "$baseUrl/api/v1$path"
        val body: RequestBody? = when {
            bodyJson != null -> bodyJson.toRequestBody(JSON_MEDIA)
            method in listOf("POST", "PUT", "PATCH") -> "{}".toRequestBody(JSON_MEDIA)
            else -> null
        }
        return Request.Builder()
            .url(url)
            .method(method, body)
            .header("Accept", "application/json")
            .header("X-Client", "mobile")
            .apply { accessToken?.let { header("Authorization", "Bearer $it") } }
            .build()
    }

    private suspend fun <T> request(path: String, method: String = "GET", bodyJson: String? = null, deserialize: (JsonElement) -> T): T =
        withContext(Dispatchers.IO) {
            val req = buildRequest(method, path, bodyJson)
            val resp = try { client.newCall(req).execute() } catch (e: Exception) { throw ApiException.NetworkError(e) }
            val bodyStr = resp.body?.string() ?: ""
            if (!resp.isSuccessful) {
                val errCode = try {
                    val env = json.parseToJsonElement(bodyStr).jsonObject
                    val err = env["error"]?.jsonObject
                    val code = err?.get("code")?.jsonPrimitive?.content ?: resp.message
                    if (resp.code == 401 && code == "unauthorized" && accessToken != null) {
                        onUnauthorized?.let { Handler(Looper.getMainLooper()).post(it) }
                        throw ApiException.Unauthorized
                    }
                    code
                } catch (e: ApiException) { throw e } catch (e: Exception) { resp.message }
                throw ApiException.HttpError(resp.code, errCode)
            }
            try {
                val env = json.parseToJsonElement(bodyStr).jsonObject
                val data = env["data"] ?: throw ApiException.DecodingError(Exception("No data in envelope"))
                deserialize(data)
            } catch (e: ApiException) { throw e } catch (e: Exception) { throw ApiException.DecodingError(e) }
        }

    private suspend fun requestVoid(path: String, method: String, bodyJson: String? = null) =
        withContext(Dispatchers.IO) {
            val req = buildRequest(method, path, bodyJson)
            val resp = try { client.newCall(req).execute() } catch (e: Exception) { throw ApiException.NetworkError(e) }
            if (resp.code == 204) return@withContext
            if (!resp.isSuccessful) {
                val bodyStr = resp.body?.string() ?: ""
                val errCode = try {
                    val env = json.parseToJsonElement(bodyStr).jsonObject
                    val err = env["error"]?.jsonObject
                    val code = err?.get("code")?.jsonPrimitive?.content ?: resp.message
                    if (resp.code == 401 && code == "unauthorized" && accessToken != null) {
                        onUnauthorized?.let { Handler(Looper.getMainLooper()).post(it) }
                        throw ApiException.Unauthorized
                    }
                    code
                } catch (e: ApiException) { throw e } catch (e: Exception) { resp.message }
                throw ApiException.HttpError(resp.code, errCode)
            }
        }

    private inline fun <reified T> decode(el: JsonElement): T = json.decodeFromJsonElement(el)
    private inline fun <reified T> decodeNullable(el: JsonElement): T? =
        if (el is JsonNull) null else json.decodeFromJsonElement(el)

    // MARK: - Auth

    suspend fun login(email: String, password: String, mfaCode: String? = null): AuthResponse {
        val body = buildJsonObject {
            put("email", email); put("password", password)
            mfaCode?.let { put("mfa_code", it) }
        }
        return request("/auth/login", "POST", body.toString()) { decode(it) }
    }

    suspend fun register(email: String, password: String, displayName: String, mekSalt: String, mekEnvelope: String): AuthResponse {
        val body = buildJsonObject {
            put("email", email); put("password", password); put("display_name", displayName)
            put("mek_salt", mekSalt); put("mek_envelope", mekEnvelope)
        }
        return request("/auth/register", "POST", body.toString()) { decode(it) }
    }

    suspend fun refreshToken(token: String): AuthResponse {
        val body = buildJsonObject { put("refresh_token", token) }
        return request("/auth/refresh", "POST", body.toString()) { decode(it) }
    }

    suspend fun logout(refreshToken: String) {
        val body = buildJsonObject { put("refresh_token", refreshToken) }
        requestVoid("/auth/logout", "POST", body.toString())
    }

    suspend fun forgotPassword(email: String) {
        val body = buildJsonObject { put("email", email) }
        requestVoid("/auth/forgot-password", "POST", body.toString())
    }

    suspend fun resetPassword(token: String, password: String) {
        val body = buildJsonObject { put("token", token); put("password", password) }
        requestVoid("/auth/reset-password", "POST", body.toString())
    }

    // MARK: - User

    suspend fun getMe(): User = request("/users/me") { decode(it) }

    suspend fun deleteAccount() = requestVoid("/users/me", "DELETE")

    suspend fun updateMe(displayName: String): User {
        val body = buildJsonObject { put("display_name", displayName) }
        return request("/users/me", "PATCH", body.toString()) { decode(it) }
    }

    suspend fun changeEmail(newEmail: String, currentPassword: String) {
        val body = buildJsonObject { put("new_email", newEmail); put("current_password", currentPassword) }
        requestVoid("/users/me/change-email", "POST", body.toString())
    }

    suspend fun changePassword(currentPassword: String, newPassword: String, newMekEnvelope: String) {
        val body = buildJsonObject {
            put("current_password", currentPassword); put("new_password", newPassword)
            put("new_mek_envelope", newMekEnvelope)
        }
        requestVoid("/users/me/change-password", "POST", body.toString())
    }

    suspend fun registerPushToken(token: String, platform: String) {
        val body = buildJsonObject { put("token", token); put("platform", platform) }
        requestVoid("/users/me/push-token", "POST", body.toString())
    }

    suspend fun deletePushToken() = requestVoid("/users/me/push-token", "DELETE")

    // MARK: - Sessions

    suspend fun listSessions(): List<Session> = request("/users/me/sessions") { decode(it) }
    suspend fun revokeSession(id: String) = requestVoid("/users/me/sessions/$id", "DELETE")
    suspend fun revokeAllSessions() = requestVoid("/users/me/sessions", "DELETE")

    // MARK: - Vaults

    suspend fun listVaults(): List<Vault> = request("/vaults") { decode(it) }

    suspend fun createVault(name: String, icon: String, cekEnvelope: String): Vault {
        val body = buildJsonObject { put("name", name); put("icon", icon); put("cek_envelope", cekEnvelope) }
        return request("/vaults", "POST", body.toString()) { decode(it) }
    }

    suspend fun updateVault(id: String, name: String? = null, icon: String? = null, accessMode: String? = null, cascadeWindowDays: Int? = null, notifyLockedTiers: Boolean? = null): Vault {
        val body = buildJsonObject {
            name?.let { put("name", it) }; icon?.let { put("icon", it) }
            accessMode?.let { put("access_mode", it) }
            cascadeWindowDays?.let { put("cascade_window_days", it) }
            notifyLockedTiers?.let { put("notify_locked_tiers", it) }
        }
        return request("/vaults/$id", "PATCH", body.toString()) { decode(it) }
    }

    suspend fun deleteVault(id: String) = requestVoid("/vaults/$id", "DELETE")

    // MARK: - Entries

    suspend fun listEntries(vaultId: String): List<VaultEntry> =
        request("/vaults/$vaultId/entries") { decode(it) }

    suspend fun createEntry(vaultId: String, entryType: String, title: String, encryptedData: String, sortOrder: Int? = null): VaultEntry {
        val body = buildJsonObject {
            put("entry_type", entryType); put("title", title); put("encrypted_data", encryptedData)
            sortOrder?.let { put("sort_order", it) }
        }
        return request("/vaults/$vaultId/entries", "POST", body.toString()) { decode(it) }
    }

    suspend fun updateEntry(vaultId: String, entryId: String, title: String? = null, encryptedData: String? = null, sortOrder: Int? = null, isFavorite: Boolean? = null): VaultEntry {
        val body = buildJsonObject {
            title?.let { put("title", it) }; encryptedData?.let { put("encrypted_data", it) }
            sortOrder?.let { put("sort_order", it) }; isFavorite?.let { put("is_favorite", it) }
        }
        return request("/vaults/$vaultId/entries/$entryId", "PATCH", body.toString()) { decode(it) }
    }

    suspend fun deleteEntry(vaultId: String, entryId: String) =
        requestVoid("/vaults/$vaultId/entries/$entryId", "DELETE")

    // MARK: - Vault Beneficiaries

    suspend fun getVaultBeneficiaries(vaultId: String): List<VaultBeneficiary> =
        request("/vaults/$vaultId/beneficiaries") { decode(it) }

    suspend fun assignBeneficiary(vaultId: String, beneficiaryId: String, cekEnvelope: String) {
        val body = buildJsonObject {
            put("beneficiary_id", beneficiaryId); put("beneficiary_cek_envelope", cekEnvelope)
        }
        requestVoid("/vaults/$vaultId/beneficiaries", "POST", body.toString())
    }

    suspend fun removeVaultBeneficiary(vaultId: String, beneficiaryId: String) =
        requestVoid("/vaults/$vaultId/beneficiaries/$beneficiaryId", "DELETE")

    suspend fun setBeneficiaryTier(vaultId: String, beneficiaryId: String, tier: String?) {
        val body = buildJsonObject { put("tier", tier?.let { JsonPrimitive(it) } ?: JsonNull) }
        requestVoid("/vaults/$vaultId/beneficiaries/$beneficiaryId/tier", "PATCH", body.toString())
    }

    // MARK: - Beneficiaries

    suspend fun listBeneficiaries(): List<Beneficiary> = request("/beneficiaries") { decode(it) }

    suspend fun createBeneficiary(name: String, email: String, relationship: String? = null, secretQuestion: String? = null, photoData: String? = null): Beneficiary {
        val body = buildJsonObject {
            put("name", name); put("email", email)
            relationship?.let { put("relationship", it) }
            secretQuestion?.let { put("secret_question", it) }
            photoData?.let { put("photo_data", it) }
        }
        return request("/beneficiaries", "POST", body.toString()) { decode(it) }
    }

    suspend fun updateBeneficiary(id: String, name: String? = null, email: String? = null, relationship: String? = null, secretQuestion: String? = null, photoData: String? = null): Beneficiary {
        val body = buildJsonObject {
            name?.let { put("name", it) }; email?.let { put("email", it) }
            relationship?.let { put("relationship", it) }
            secretQuestion?.let { put("secret_question", it) }
            photoData?.let { put("photo_data", it) }
        }
        return request("/beneficiaries/$id", "PATCH", body.toString()) { decode(it) }
    }

    suspend fun deleteBeneficiary(id: String) = requestVoid("/beneficiaries/$id", "DELETE")

    suspend fun resendBeneficiaryConfirmation(id: String) =
        requestVoid("/beneficiaries/$id/resend", "POST")

    suspend fun getBeneficiaryVaults(id: String): List<BeneficiaryVaultItem> =
        request("/beneficiaries/$id/vaults") { decode(it) }

    // MARK: - Switch

    suspend fun getSwitchSettings(): SwitchSettings = request("/switch") { decode(it) }

    suspend fun updateSwitchSettings(
        isActive: Boolean? = null,
        checkInIntervalDays: Int? = null,
        abortWindowHours: Int? = null,
        reminder1DaysBefore: Int? = null,
        reminder2HoursBefore: Int? = null,
        finalWarningHoursBefore: Int? = null,
        preferredCheckinHour: Int? = null,
        clearPreferredHour: Boolean? = null,
        timezone: String? = null
    ): SwitchSettings {
        val body = buildJsonObject {
            isActive?.let { put("is_active", it) }
            checkInIntervalDays?.let { put("check_in_interval_days", it) }
            abortWindowHours?.let { put("abort_window_hours", it) }
            reminder1DaysBefore?.let { put("reminder1_days_before", it) }
            reminder2HoursBefore?.let { put("reminder2_hours_before", it) }
            finalWarningHoursBefore?.let { put("final_warning_hours_before", it) }
            preferredCheckinHour?.let { put("preferred_checkin_hour", it) }
            clearPreferredHour?.let { put("clear_preferred_hour", it) }
            timezone?.let { put("timezone", it) }
        }
        return request("/switch", "PATCH", body.toString()) { decode(it) }
    }

    suspend fun checkin(): SwitchSettings = request("/switch/checkin", "POST") { decode(it) }

    suspend fun pauseSwitch(resumeAt: String? = null): SwitchSettings {
        val body = buildJsonObject { resumeAt?.let { put("resume_at", it) } }
        return request("/switch/pause", "POST", body.toString()) { decode(it) }
    }

    suspend fun resumeSwitch(): SwitchSettings = request("/switch/resume", "POST") { decode(it) }

    suspend fun abortTrigger(): SwitchSettings = request("/switch/abort", "POST") { decode(it) }

    suspend fun revokeDeliveries() = requestVoid("/switch/revoke-deliveries", "POST")

    // MARK: - Trusted Contacts

    suspend fun listTrustedContacts(): List<TrustedContact> =
        request("/trusted-contacts") { decode(it) }

    suspend fun createTrustedContact(
        name: String, email: String, phone: String? = null,
        notifyOnFinalWarning: Boolean = false, canAbort: Boolean = false,
        canVerifyLife: Boolean = false, canCorroborateDeath: Boolean = false
    ): TrustedContact {
        val body = buildJsonObject {
            put("name", name); put("email", email); phone?.let { put("phone", it) }
            put("notify_on_final_warning", notifyOnFinalWarning); put("can_abort", canAbort)
            put("can_verify_life", canVerifyLife); put("can_corroborate_death", canCorroborateDeath)
        }
        return request("/trusted-contacts", "POST", body.toString()) { decode(it) }
    }

    suspend fun updateTrustedContact(
        id: String, name: String? = null, phone: String? = null,
        notifyOnFinalWarning: Boolean? = null, canAbort: Boolean? = null,
        canVerifyLife: Boolean? = null, canCorroborateDeath: Boolean? = null
    ): TrustedContact {
        val body = buildJsonObject {
            name?.let { put("name", it) }; phone?.let { put("phone", it) }
            notifyOnFinalWarning?.let { put("notify_on_final_warning", it) }
            canAbort?.let { put("can_abort", it) }
            canVerifyLife?.let { put("can_verify_life", it) }
            canCorroborateDeath?.let { put("can_corroborate_death", it) }
        }
        return request("/trusted-contacts/$id", "PATCH", body.toString()) { decode(it) }
    }

    suspend fun deleteTrustedContact(id: String) = requestVoid("/trusted-contacts/$id", "DELETE")

    // MARK: - MFA

    suspend fun setupTOTP(): TOTPSetupResponse = request("/auth/mfa/setup", "POST") { decode(it) }

    suspend fun confirmTOTP(secret: String, code: String, backupCodes: List<String>) {
        val body = buildJsonObject {
            put("secret", secret); put("code", code)
            putJsonArray("backup_codes") { backupCodes.forEach { add(it) } }
        }
        requestVoid("/auth/mfa/verify", "POST", body.toString())
    }

    suspend fun disableMFA(code: String) {
        val body = buildJsonObject { put("code", code) }
        requestVoid("/auth/mfa/disable", "POST", body.toString())
    }

    // MARK: - Branding

    suspend fun getBranding(): BrandingResponse = request("/branding") { decode(it) }

    suspend fun updateAccentColor(hex: String) {
        val body = buildJsonObject { put("app_accent_color", hex) }
        requestVoid("/admin/config", "PATCH", body.toString())
    }

    suspend fun getAdminConfig(): Map<String, String> =
        request("/admin/config") { decode(it) }

    suspend fun submitAccessRequest(name: String, email: String, message: String) {
        val body = buildJsonObject { put("name", name); put("email", email); put("message", message) }
        requestVoid("/access-request", "POST", body.toString())
    }

    // MARK: - Death Report

    suspend fun getActiveDeathReport(): DeathReport? =
        withContext(Dispatchers.IO) {
            val req = buildRequest("GET", "/report/active")
            val resp = try { client.newCall(req).execute() } catch (e: Exception) { throw ApiException.NetworkError(e) }
            val bodyStr = resp.body?.string() ?: ""
            if (!resp.isSuccessful) return@withContext null
            try {
                val env = json.parseToJsonElement(bodyStr).jsonObject
                val data = env["data"] ?: return@withContext null
                if (data is JsonNull) return@withContext null
                json.decodeFromJsonElement<DeathReport>(data)
            } catch (e: Exception) { null }
        }
}
